### ISSUE TITLE
Meta-chords behaves inconsistently in modes

### ISSUE THREAD

**Issue by ryvnf** [neovim/neovim#10132]
### Version/platform info

- NVIM v0.3.7
- Operating system: GNU/Linux
- Terminal name: xterm

### Steps to reproduce using `nvim -u NORC`

1. Open neovim with `nvim -u NORC` in xterm
2. Press `i` to go to insert mode
3. Press Alt-o
4. Press `:` to go to command mode
5. Press Alt-o

### Actual behaviour

When pressing Alt-o in insert mode the command behaves as if `<Esc>o` was pressed when Alt-o is unmapped. It leaves insert mode and opens a new line.

When pressing Alt-o in command-line mode the command behaves as if `o` was pressed when Alt-o is unmapped. It only inserts `o`. When pressing Alt-o in visual and normal-mode it behaves like command-mode and gets treated as only `o` when Alt-o is unmapped.

### Expected behaviour

In Vim the behavior is that Alt-o is always treated as if `<Esc>o` was pressed when Alt-o is unmapped. So when pressing Alt-o command-line mode it should return to normal mode and open a new line.

A behavior which deviates from insert-mode but might be more convenient than the default Vim behavior is that Alt-*char* should be treated a no-operation unless Alt-*char* is pressed. Accidentally pressing the wrong Alt-*char* chord can be frustrating because it changes modes or aborts something.

The decision to have Alt-*char* be treated as `<Esc>char` when Alt-*char* is unmapped was taken in #8213. It seems like there were valid reasons to have it be handled that way. Because that decision has already been made I think the behavior should be changed so it gets treated like `<Esc>char` in all modes, which would make the behavior consistent and  improve compatibility with Vim.

### Why this is a problem

It is inconsistent when keycodes behaves differently from mode to mode. This can cause a lot of confusion when writing mappings. For example, the mapping `<Esc>f` can be triggered in insert-mode using Alt-F, but the mapping `<Esc>f` cannot be triggered using Alt-F in normal, command-line, or visual-mode. I ran into this when developing a plugin, and after some head-scratching and investigating, I found this issue.

**Comment by justinmk** [comment]
> the mapping `<Esc>f` can be triggered in insert-mode using Alt-F,

? Do not map `<esc>foo`, that is an accident from vim/terminal history. Map `<m-f>`.

btw, what does gvim do?

**Comment by ryvnf** [comment]
> btw, what does gvim do?

In gvim pressing Alt-*key* generates the keycode <m-key> which is encoded with the eight bit being set. This means that Alt-d inserts `ä` on my machine.

**Comment by ryvnf** [comment]
It appears that I was wrong in that the <Esc>f mapping can be triggered using Alt-F in insert-mode. <Esc>f mappings won't get triggered after pressing <M-f> and <M-f> isn't mapped, instead it gets treated like the user presses <Esc>f but ignores mappings. I think this can be considered a bug.

**Comment by justinmk** [comment]
> `<Esc>f` mappings won't get triggered after pressing `<M-f>`

Why would it? it's not the same mapping. `<esc>x` is nonsense in Nvim. If you want to use _alt_, then map to _alt_.

> instead it gets treated like the user presses `<Esc>f` but ignores mappings. I think this can be considered a bug.

No, that doesn't make sense. I tried to explain this  in 
https://github.com/neovim/neovim/issues/10132#issuecomment-499546915 . Just because Vim accidentally behaved like this does not mean this behavior should be kept. 

Let's focus on the enhancement request in this ticket: making the "fall through" behavior work for modes other than insert-mode. 

Making `<esc>x` mappings behave like `<alt-x>` for mappings is `wontfix`. This is not a "feature" that serves any purpose. If you want this feature, use Vim.

**Comment by ryvnf** [comment]
Yes! I know the common practice is to map <M-f> instead of <Esc>f in Nvim, and for good reason.

Reason I thought it could be a bug is that the reason issue #8213 for introducing the behavior was that it is possible to confuse Nvim when pressing <Esc> followed by another key rapidly, because Nvim thinks it was part of a meta mapping. Then the behavior to make it behave like <Esc> followed by another key-press was introduced. But it seems weird to treat it like <Esc>f but not make it trigger mappings.

Someone could be in the same situation as the author of #8213 but having a mapping <esc>x which won't trigger if the author presses it quickly. <esc>x might to be mapped without the intention of it being activated with the ALT-modifier. But it won't work if pressing the keys rapidly like the author of #8213. It isn't common to map escape like this so this probably doesn't affect many people (if anyone).

**Comment by justinmk** [comment]
> Reason I thought it could be a bug is that the reason issue #8213 for introducing the behavior was that it is possible to confuse Nvim when pressing `<Esc>` followed by another key rapidly, because Nvim thinks it was part of a meta mapping. 

No. #8213 was strictly an enhancement to match an old Vim behavior in the _unmapped_ case. 

Terminal applications being confused by rapid `<esc> + x` is unavoidable (can be tuned with `:help 'ttimeout'` and tmux `escape-time`).

> But it seems weird to treat it like `<Esc>f` but not make it trigger mappings.

If `<a-x>` is mapped then the user explicitly opted-into interpreting it as ALT rather than ESC. It is impossible to disambiguate the two. When it is _not mapped_ then there is no ambiguity because `<a-x>` by default does nothing. That's why #8213 was feasible.

**Comment by godlygeek** [comment]
I've just started trying out neovim for the first time, and this has been driving me absolutely crazy - no matter how I tweak terminal timeouts, I occasionally hit another character fast enough after `<Esc>` that both the mode change and the next command are swallowed silently.  I've hit this repeatedly trying to exit visual mode and do something else.

I've entered a PR that should apply the behavior that was adopted for insert mode in #8213 in all modes.
